src/PolyType/Abstractions/IDictionaryTypeShape.cs

```diff
@@ -74,9 +74,10 @@ public interface IDictionaryTypeShape<TDictionary, TKey, TValue> : ITypeShape<TD
     /// <summary>
     /// Creates a delegate wrapping a parameterless constructor of a mutable collection.
     /// </summary>
+    /// <param name="collectionConstructionOptions">Optional options that control how the collection is constructed.</param>
     /// <exception cref="InvalidOperationException">The collection is not <see cref="CollectionConstructionStrategy.Mutable"/>.</exception>
     /// <returns>A delegate wrapping a default constructor.</returns>
-    Func<TDictionary> GetDefaultConstructor();
+    Func<TDictionary> GetDefaultConstructor(in CollectionConstructionOptions<TKey> collectionConstructionOptions = default);
```

Comment:
Should we consider building in capacity support to avoid future breaking changes? If it makes the diff too complicated, we could just leave it as an a no-op in the shape providers.
```suggestion
    Func<int?, TDictionary> GetDefaultConstructor(in CollectionConstructionOptions<TKey> collectionConstructionOptions = default);
```

Reply:
I think you mentioned elsewhere the (potentially over-engineered) possibility of taking a custom struct as input instead of just `int?`. If we think we _might_ support allowing comparers to be provided at activation time instead of visitor-time, a custom struct approach might defend against that future possibility. Thoughts?
I'm leaning toward saving this for a subsequent change. I like the idea of front-loading it to reduce breaking changes on the roadmap, but I'd also like to get _this_ comparer change in ASAP so that we can keep building on it within the higher-level libraries.

Reply:
> If we think we _might_ support allowing comparers to be provided at activation time instead of visitor-time, a custom struct approach might defend against that future possibility. Thoughts?

While I can't discount the possibility of this ever coming up as a requirement, it seems unlikely (payloads describing their own equality comparison? Sounds kind of dangerous). That said, maybe I'm overestimating the performance penalty incurred by the ctor delegate having a bit more branching than strictly necessary. Maybe we should just keep things simple and have one options struct accepted by the delegate.

Reply:
Regarding the conditional branching, here's another idea: what if instead of a method that returns just one delegate, it returns a struct of delegates. One delegates takes an `int` and one takes no args. That way, no branching is necessary at any level, unless a converter can't predict whether a collection will be length prefixed or not, and will have to branch anyway based on that.
By returning a struct of delegates instead of adding new ITypeShape members that return one delegate each, we allow for backward compatibility when adding new delegates.

Reply:
In the case of the reflection provider, wouldn't this require that all delegates be generated even if they're not being used?

Reply:
Yes. But I suppose we can mitigate that by returning a _class_ of delegates instead of a struct of delegates. If the class defined virtual property getters for the delegates, the reflection providers can override the getters to generate the delegates on-demand.

Reply:
I think ultimately this creates a shape type within a shape type. It could be exposed as a separate factory directly on shape type.

Reply:
But shapes are _interfaces_ which makes adding members a breaking change. My proposal allows adding members that _don't_ require API breaking changes.

Reply:
Yeah.. I should get to changing those to abstracts at some point, at least the generic ones.

Reply:
> It could be exposed as a separate factory directly on shape type.

If this is the way you want to go, does that mean we'll double up like this, for each existing factory?

```cs
Func<TDictionary> GetDefaultConstructor(in CollectionConstructionOptions<TKey> collectionConstructionOptions = default);
Func<int, TDictionary> GetDefaultConstructorWithCapacity(in CollectionConstructionOptions<TKey> collectionConstructionOptions = default);
```

In any case, I'd ask that capacity factories be held off for a subsequent PR.

Reply:
> In any case, I'd ask that capacity factories be held off for a subsequent PR.

Fine by me.